### DISCUSSION TITLE
How to automate based on an ever-changing time?

### DISCUSSION THREAD

**Discussion by wsw70** [custom-components/pyscript#186] [Q&A] [answered]
Context: I have roller shutters on windows that I can automate. Today they go up and down based on a set time (or an offset from sunset). I would like to make them go up based on a time that is reassessed every minute.

Details: I retrieve the school start hours of my children from the school site and they are available as MQTT messages, updated every minute or so. I would like to move the shutters up an hour before school starts.

The problem: it should be a one-time "up" per day. If someone wants to pull them down I do not want them to go up again just because it is "later than an hour to school".

Before I get into a Rube Goldberg machine kind of automation I am asking, in case someone has already thought this out and has a reasonable approach

**Comment by craigbarratt** [comment, marked as answer]
Here's one approach.  Just trigger off the MQTT message, and the task does a `task.unique` and then a `task.wait_until` until an hour before the school time (or skip if it's already less than an hour).  After the `task.wait_until` it opens the shades.  Each time you get a new MQTT update, the old task gets killed and replaced by the new one, which uses the potentially new time to do the next `task.wait_until`.

To make this only happen once per day, you'll also have to have a flag (eg, some state variable) that is set when the shades are auto-opened.  Once the flag is set, the MQTT task trigger should be disabled by putting that condition in `@state.active`.  You can reset the flag each morning with a fixed time trigger at, eg, 5am.

**Reply by wsw70** [reply]
In the meantime I ended up having a persistent flag (you suggested that as well) and just check for conditions:

```
# setting persistant stae (and creating if needed)
for who in ['martin', 'michael']:
    var = f'pyscript.volets_{who}_up'
    try:
        state.get(var)
    except NameError:
        state.set(var, False)
    state.persist(var)

@mqtt_trigger('edt')
def edt(**data):
    payload = json.loads(data['payload'])
    for who in ['martin', 'michael']:
        if time := payload[f'{who}{"_matin" if who=="martin" else ""}']['today']['start']:
            if state.get(f'pyscript.volets_{who}_up'):
                # volets already up
                log.info(f"volets {who} already up")
                pass
            else:
                time = arrow.get(time)
                if time.shift(hours=+1) > arrow.now():
                    # move the rollers up
                    cover.open_cover(entity_id=f"cover.{who}_volets")
                    state.set(f'pyscript.volets_{who}_up', True)
                    log.info(f"move {who} volets up, time to wake up for school")
        else:
            # pas de cours
            log.info(f"volets {who} pas de cours")
            pass

@time_trigger("once(00:01)")
def reset_volets_persistent():
    for who in ['martin', 'michael']:
        var = f'pyscript.volets_{who}_up'
        state.set(var, False)
```

`edt` is triggered at each MQTT message, it checks if there is a school time (vs. week-ends or vacation), then if the shutters are already up (the flag), and finally moving them up and setting the flag.
The flag is reset at midnight.

This is not completely foolproof but should be good enough if the school timetable does not go crazy (this is France so you never know...) 

Sorry for the weird comments, I mixed up French and English

**Reply by craigbarratt** [reply]
That's a simpler way to do it than I suggested.  So long as the MQTT events arrive regularly enough, your approach of simply raising the shades the first time it's less than an hour away looks great.

My approach of using `task.wait_until` would be necessary if there were only a small number of MQTT events well before the action is needed, since that method waits until that time, so it doesn't need a new MQTT event just after the desired time.

**Reply by wsw70** [reply]
> My approach of using task.wait_until would be necessary if there were only a small number of MQTT events well before the action is needed, since that method waits until that time, so it doesn't need a new MQTT event just after the desired time.

Yes, that's a good point. My MQTT messages arrive every few minutes so that should be fine.

**Reply by dorbitbrown** [reply]
Looks nice.  I'm doing something similar.  Part of my use case is that during daylight I calculate based on sun.sun's azimuth and elevation  and models of the horizon and my home's eaves whether there is direct sunlight coming in any of 10 windows, and then position shades to prevent sun damage to floors and furnishings. 

Like you, I needed a way that the users could manually (or using non-HA compatible remote controls) reposition shades to override the automatic repositioning, creating a "temporary hold" state that continued until the next time-of-day state change (dawn, day, dusk, night).  I did this by polling (via HTTP GET) the Hunter Douglas Powerview hub for the shades' positions and if they weren't where the automation routine had last positioned them, taking that as a command to stop moving them automatically.  In this case the shade's control_state attribute would be set to 'manual', and move commands for that shade ignored until the state later got reset to 'auto' at the next time-of-day state change.

Experience soon proved it would be nice to be able to cancel the hold state using a remote, without having to get into the HA dashboard.  Fortunately the remotes allow definition of a favorite position for each shade, so I repurposed those positions to be a signal that automatic repositioning should resume. So, if a polled shade is found to be in its favorite position, the automatic repositioning function changes the control_state attribute for that shade back to 'auto'.
